src/psum.c

```diff
@@ -0,0 +1,1018 @@
+#include "data.table.h"
+
+// like base::p{max,min}, but for sum
+SEXP psum(SEXP x, SEXP narmArg) {
+  if (!isNewList(x)) error(_("Internal error: x must be a list"));
+  if (!isLogical(narmArg) || LENGTH(narmArg)!=1 || LOGICAL(narmArg)[0]==NA_LOGICAL) error(_("na.rm must be TRUE or FALSE"));
+
+  int J=LENGTH(x);
+  if (J == 0) {
+    error(_("Empty input"));
+  } else if (J == 1) {
+    SEXP xj0 = VECTOR_ELT(x, 0);
+    if (TYPEOF(xj0) == VECSXP) { // e.g. psum(.SD)
+      x = duplicate(xj0);
+      J = LENGTH(xj0);
+    } else {
+      // na.rm doesn't matter -- input --> output
+      return duplicate(xj0);
+    }
+  }
+  SEXPTYPE outtype = INTSXP;
+  int n = -1;
+  for (int j=0; j<J; j++) {
+    SEXP xj=VECTOR_ELT(x, j);
+    switch(TYPEOF(xj)) {
+    case LGLSXP: case INTSXP:
+      if (isFactor(xj)) {
+        error(_("psum not meaningful for factors"));
+      }
+      break;
+    case REALSXP:
+      if (INHERITS(xj, char_integer64)) {
+        error(_("integer64 input not supported"));
+      }
+      if (outtype == INTSXP) { // bump if this is the first numeric we've seen
+        outtype = REALSXP;
+      }
+      break;
+    case CPLXSXP:
+      if (outtype != CPLXSXP) { // only bump if we're not already complex
+        outtype = CPLXSXP;
+      }
+      break;
+    default:
+      error(_("Only logical, numeric & complex inputs are supported for %s"), "psum");
+    }
+    if (n >= 0) {
+      int nj = LENGTH(xj);
+      if (n == 1 && nj > 1) {
+        n = nj; // singleton comes first, vector comes later [psum(1, 1:4)]
+      } else if (nj != 1 && nj != n) {
+        error(_("Inconsistent input lengths -- first found %d, but %d element has length %d. Only singletons will be recycled."), n, j+1, nj);
+      }
+    } else { // initialize
+      n = LENGTH(xj);
+    }
+  }
+
+  SEXP out = PROTECT(allocVector(outtype, n));
+  if (n == 0) {
+    UNPROTECT(1);
+    return(out);
+  }
+  if (LOGICAL(narmArg)[0]) {
+    // initialize to NA to facilitate all-NA rows --> NA output
+    writeNA(out, 0, n);
+    switch (outtype) {
+    case INTSXP: {
+      int *outp = INTEGER(out);
+      for (int j=0; j<J; j++) {
+        SEXP xj = VECTOR_ELT(x, j);
+        int nj = LENGTH(xj);
+        int *xjp = INTEGER(xj); // INTEGER is the same as LOGICAL
+        for (int i=0; i<n; i++) {
+          int xi = nj == 1 ? 0 : i; // recycling for singletons
+          if (xjp[xi] != NA_INTEGER) { // NA_LOGICAL is the same
+            if (outp[i] == NA_INTEGER) {
+              outp[i] = xjp[xi];
+            } else {
+              if (INT_MAX - xjp[xi] < outp[i] || INT_MIN - xjp[xi] < outp[i]) { // overflow
```

Comment:
INT_MAX - negative value overflows, I think.

Reply:
will fix this when i write tests or I'm sure I'll forget to write the test